[serve] Add experimental support for `StreamingResponse` using `RayObjectRefGenerator` (#35720)
ray-project/ray: doc/source/serve/http-guide.md

Serve currently does not support WebSockets. If you have a use case that requires it, please [let us know](https://github.com/ray-project/ray/issues/new/choose)!
```

> (serve-http-streaming-response)=

Reviewer: Should we pull this in a separate page for discoverability (or rename this to something better than HTTP Guide)?

Author: We already have a lot of user guides and this feels like the natural place for it (where we explain starlette & fastapi behaviors)

Author: no preference on the naming if you have a better one suggest it

Author: Once we add batching + a real-world example we can have a separate page for that. Not enough time before branch cut for those though

Reviewer: That sounds good, we have more time for docs changes btw, so we can do it in the follow-up PR.

Author: Ah ok I'll work on it first thing next week then

## Streaming Responses

```{warning}

> Support for HTTP streaming responses is currently experimental. To enable this feature, set `RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING=1 on the cluster before starting Ray. If you encounter any issues, please [file an issue on GitHub](https://github.com/ray-project/ray/issues/new/choose).

Reviewer 2: ```suggestion
Support for HTTP streaming responses is experimental. To enable this feature, set `RAY_SERVE_ENABLE_EXPERIMENTAL_STREAMING=1` on the cluster before starting Ray. If you encounter any issues, [file an issue on GitHub](https://github.com/ray-project/ray/issues/new/choose).
```

```

> Some applications require streaming incremental results back to the caller.

Reviewer 2: ```suggestion
Some applications must stream incremental results back to the caller.
```

> This is common for large language models (LLMs) used for text generation.

Reviewer: or video processing?

Author: added

> This is common for large language models (LLMs) used for text generation or video processing applications.

Reviewer 2: ```suggestion
This is common for text generation or video processing applications that use large language models (LLMs).
```

The full forward pass may take multiple seconds, so providing incremental results as they're available provides a much better user experience.

> To use HTTP response streaming, return a [StreamingResponse](https://www.starlette.io/responses/#streamingresponse) that wraps a generator from your HTTP handler.

Reviewer: Also include setting the env variable and how you would do it in yaml.

Author: Set the env variable in cluster env

Reviewer: Got it, definitely need to improve in 2.6. Let's document in Anyscale docs too once you're done with the ray changes.

This is supported for basic HTTP ingress deployments using a `__call__` method and when using the [FastAPI integration](serve-fastapi-http).

The code below defines a Serve application that incrementally streams numbers up to a provided `max`.
The client-side code is also updated to handle the streaming outputs.

> In this case, we are using the `stream=True` option to the [requests](https://requests.readthedocs.io/en/latest/user/advanced/#streaming-requests) library.

Reviewer 2: ```suggestion
This case uses the `stream=True` option to the [requests](https://requests.readthedocs.io/en/latest/user/advanced/#streaming-requests) library.
```
